### PR TITLE
add handler on delete to ignore read timeout

### PR DESCRIPTION
**Pull request by scefali**
Currently, if we get a `ReadTimeout` when uninstalling a Sentry app, the uninstallation fails. This can happen if the server never responds to the webhook. This can cause an issue because users cannot uninstall broken integrations.

### CODE REVIEW

File: tests/sentry/mediators/sentry_app_installations/test_destroyer.py [hunk 1]
```diff
@@ -46,6 +47,19 @@ def test_deletes_grant(self):
...
         assert not ApiGrant.objects.filter(pk=grant.id).exists()
 
+    @responses.activate
+    def test_deletes_on_read_timeout_error(self):
+        grant = self.install.api_grant
+
+        responses.add(
+            responses.POST,
+            'https://example.com/webhook',
+            body=ReadTimeout('Read timed out'))
```
Comment (mnoble) [inline tests/sentry/mediators/sentry_app_installations/test_destroyer.py]: Does this work with `responses`? Like, does this result in receiving a 408 or whatever?

Comment (scefali) [reply tests/sentry/mediators/sentry_app_installations/test_destroyer.py]: @mnoble Not sure I quite understand your question. But there seems to be a difference between getting a status code that is bad (e.g. 500) and a run time exception like a `ReadTimeout` where there is no status code.

Comment (mnoble) [reply tests/sentry/mediators/sentry_app_installations/test_destroyer.py]: oh I see. So our http lib would throw a `ReadTimeout` exception in this case. I guess then my question is, does `responses` do that when you set a `body` like this.

Normally, `body` is meant to be the string value of the response body. Setting it to an instance of an object is counter-intuitive to me, hence the question.

Comment (scefali) [reply tests/sentry/mediators/sentry_app_installations/test_destroyer.py]: @mnoble OK, I understand now. While it does seem weird, the docs for `responses` say you should throw errors like this: https://github.com/getsentry/responses

> Lastly, you can pass an Exception as the body to trigger an error on the request:

```
import responses
import requests

@responses.activate
def test_simple():
    responses.add(responses.GET, 'http://twitter.com/api/1/foobar',
                  body=Exception('...'))
    with pytest.raises(Exception):
        requests.get('http://twitter.com/api/1/foobar')
```

Does that answer your question?

File: src/sentry/mediators/sentry_app_installations/destroyer.py [hunk 2]
```diff
@@ -34,11 +35,16 @@ def _destroy_service_hooks(self):
...
-                action='deleted',
-            )
+            try:
+                InstallationNotifier.run(
+                    install=self.install,
+                    user=self.user,
+                    action='deleted',
+                )
+            except ReadTimeout:
+                # if we get a read timeout, just move on and delete the installation
```
Comment (markstory) [inline src/sentry/mediators/sentry_app_installations/destroyer.py]: Do you want logging or metrics on this?

Comment (scefali) [reply src/sentry/mediators/sentry_app_installations/destroyer.py]: @markstory We probably should have logs for any error. Is that not automatically captured at the request level?

File: tests/sentry/mediators/sentry_app_installations/test_destroyer.py (hunk 1 above)
Comment (mnoble) [reply tests/sentry/mediators/sentry_app_installations/test_destroyer.py]: TIL. Yup, that does. Thanks!

File: src/sentry/mediators/sentry_app_installations/destroyer.py (hunk 2 above)
Comment (mnoble) [reply src/sentry/mediators/sentry_app_installations/destroyer.py]: Not if you catch it and `pass`. The SDK never gets a chance to handle the error in this case. We can add:

```python
from sentry_sdk import capture_exception

...

except ReadTimeout as e:
    capture_exception(e)

...
```

though and it should send it along to Sentry.

**Comment by scefali** [comment]
@markstory @mnoble Added the `capture_exception`. I also added a case to handle a `ConnectionError` since that could happen if the webhook server is down. I am wondering if I need to handle other types of network errors as well like `SSLError`. Please advise.

File: src/sentry/mediators/sentry_app_installations/destroyer.py (hunk 2 above)
Comment (markstory) [reply src/sentry/mediators/sentry_app_installations/destroyer.py]: We capture 500s in the internal sentry project. However, if an integration partner is having downtime/timeout conditions do we need to be alerted?

File: src/sentry/mediators/sentry_app_installations/destroyer.py [hunk 3]
```diff
@@ -34,11 +36,19 @@ def _destroy_service_hooks(self):
...
-            )
+            try:
+                InstallationNotifier.run(
+                    install=self.install,
+                    user=self.user,
+                    action='deleted',
+                )
+            except Exception as exc:
+                capture_exception(exc)
+                # if the timeout is not due to timeout or connectivity, then raise it to
```
Comment (MeredithAnya) [inline src/sentry/mediators/sentry_app_installations/destroyer.py]: bb nitpick I think you meant `if the exception is not due to...`

Comment (scefali) [reply src/sentry/mediators/sentry_app_installations/destroyer.py]: @MeredithAnya fixed

**Comment by scefali** [comment]
@mnoble logging has been added

File: src/sentry/mediators/sentry_app_installations/destroyer.py (hunk 2 above)
Comment (mnoble) [reply src/sentry/mediators/sentry_app_installations/destroyer.py]: I'd say yes, if we have some way of doing something in that scenario, but as we are today, I don't think we do. I think it's worth having the data, though.

**Review by mnoble — APPROVED** [review]
👏

**Pull request merged**